### PR TITLE
feat: add notify telegram & bark, and some minor fixes

### PR DESCRIPTION
**Pull request by lroolle**
- Add send telegram bot example;
- Add notify bark example;
- Reduce KV writes by skipping the good status, and support a `kvWriteCooldownMinutes` settings; thus we can check the status every 1 minutes in the workers free-tire plan;
- Add support to configure `url` for href in the name element;

### CODE REVIEW

**Comment by lyc8503** [comment]
Thank you very much for your pull request, let me test it.

File: worker/src/index.ts [hunk 1]
```diff
@@ -58,8 +58,12 @@ export default {
     state.overallDown = 0
     state.overallUp = 0
 
+    let statusChaged = false
```
Comment (lyc8503) [inline worker/src/index.ts]: You made a typo here, it should be `statusChanged`. So your reference to `statusChanged` below actually refers to an undefined variable.

File: .github/workflows/deploy.yml [hunk 2]
```diff
@@ -56,25 +61,26 @@ jobs:
...
         continue-on-error: true
-        
+
         run: |
           terraform init
           terraform apply -auto-approve -input=false
         env:
           CLOUDFLARE_API_TOKEN: ${{ secrets.CLOUDFLARE_API_TOKEN }}
-          TF_VAR_CLOUDFLARE_ACCOUNT_ID: ${{ steps.fetch_account_id.outputs.account_id }}
+          TF_VAR_CLOUDFLARE_ACCOUNT_ID: ${{ secrets.CLOUDFLARE_ACCOUNT_ID || steps.fetch_account_id.outputs.account_id }}
```
Comment (lyc8503) [inline .github/workflows/deploy.yml]: Maybe `secrets.CLOUDFLARE_ACCOUNT_ID ||` is not neccessary here. You have set `fetch_account_id.outputs.account_id` to `secrets.CLOUDFLARE_ACCOUNT_ID` above?

**Comment by lyc8503** [comment]
Another concern I have is that if I skip updating the state stored in the kv, the update time displayed on the front end will not be updated in time.

I've been writing code under the assumption that the interval between each update is fixed, and I may need to double-check that there are no other bugs caused by kvWriteCooldown.

Thanks a lot for the PR, I might be a bit busy these days, I'll merge in time after testing to make sure there are no more issues.

**Comment by lroolle** [comment]
> Another concern I have is that if I skip updating the state stored in the kv, the update time displayed on the front end will not be updated in time.
> 
> I've been writing code under the assumption that the interval between each update is fixed, and I may need to double-check that there are no other bugs caused by kvWriteCooldown.

 

1. I believe the update time of the status page is not that important, especially after adding the notify mechanism, you'll find that timeliness is actually more crucial.
2. The last update time displayed on the page doesn't seem to conflict with the actual last check time. If the cooldown is set too high, the displayed data points will be a bit sparse.
3. I set the cooldown to 10 minutes, and it has been running for about five days after the actual deployment. So far, there doesn't seem to be any problems 😁. Longer durations have not been tested, so it may be necessary to limit the maximum value of the cooldown.
4. Also, I'm not sure which module is best for placing the code. Ideally, it should be possible to use Telegram or Bark and similar services by simply adding a few configurations.

**Comment by lroolle** [comment]
@lyc8503 It seems like I do not have the privilege to update the wiki page, FYI the guides are pasted below:

### Optionally: Setup Telegram Bot

To setup your own Telegram Bot:

1. Create a bot with BotFather at https://t.me/botfather, and obtain the bot token.
2. Create a group chat/Channel and add the bot to the group.
3. Obtain the chat ID, you can:
    - By adding https://t.me/raw_data_bot to your chanel or group; (*Recommended*)
    - Or by sending a message to the bot and then visiting https://api.telegram.org/bot<YourBOTToken>/getUpdates; [Read More](https://gist.github.com/nafiesl/4ad622f344cd1dc3bb1ecbe468ff9f8a)
4. Set you worker environment variables for:
   - SECRET_TELEGRAM_CHAT_ID: string;
   - SECRET_TELEGRAM_API_TOKEN: string;


### Optionally: Setup [Bark](https://bark.day.app/)

To setup your own Bark:

1. Follow the instructions on https://bark.day.app/#/deploy (*onrender is recommended*)
2. Set you worker environment variables for:
  - BARK_SERVER: string;
  - BARK_DEVICE_KEY: string;

**Comment by lyc8503** [comment]
> Also, I'm not sure which module is best for placing the code. Ideally, it should be possible to use Telegram or Bark and similar services by simply adding a few configurations.

You have a point, but the logic of pushing messages can be complicated, for example, I want to skip downtime reminders that are too short (because they are more likely to be false positive), and the development of adapting to everyone's needs can be even more complicated, so I'm using callbacks for the time being, so that each person can write the code according to their own needs.

**Comment by lyc8503** [comment]
I have fixed some more minor bugs and updated wiki and online demo. Thanks for your contribution!

**Pull request merged**